snapshot_iterator.go

```diff
@@ -0,0 +1,301 @@
+)
+
+type (
+	snapshotIterator struct {
+		t            *tomb.Tomb
+		acks         csync.WaitGroup
+		lastPosition common.SnapshotPosition
+		config       snapshotIteratorConfig
+		dataC        chan fetchData
+	}
+	snapshotIteratorConfig struct {
+		tableKeys common.TableKeys
+		client    *spanner.Client
+	}
+	fetchData struct {
+		payload        opencdc.StructuredData
+		table          common.TableName
+		primaryKeyName common.PrimaryKeyName
+		primaryKeyVal  any
+		position       common.TablePosition
+	}
+)
+
+var _ common.Iterator = new(snapshotIterator)
+
+var ErrSnapshotIteratorDone = errors.New("snapshot complete")
+
+func newSnapshotIterator(ctx context.Context, config snapshotIteratorConfig) *snapshotIterator {
+	t, _ := tomb.WithContext(ctx)
+	iterator := &snapshotIterator{
+		t:      t,
+		acks:   csync.WaitGroup{},
+		config: config,
+		dataC:  make(chan fetchData),
+		lastPosition: common.SnapshotPosition{
+			Snapshots: map[common.TableName]common.TablePosition{},
+		},
+	}
+
+	for tableName, primaryKey := range config.tableKeys {
+		iterator.t.Go(func() error {
+			return iterator.fetchTable(ctx, tableName, primaryKey)
+		})
+	}
+
+	return iterator
+}
+
+func (s *snapshotIterator) Read(ctx context.Context) (rec opencdc.Record, err error) {
+	select {
+	case <-ctx.Done():
+		//nolint:wrapcheck // no need to wrap canceled error
+		return rec, ctx.Err()
+	case <-s.t.Dead():
+		if err := s.t.Err(); err != nil && !errors.Is(err, ErrSnapshotIteratorDone) {
+			return rec, fmt.Errorf(
+				"cannot stop snapshot mode, fetchers exited unexpectedly: %w", err)
+		}
+		if err := s.acks.Wait(ctx); err != nil {
+			return rec, fmt.Errorf("failed to wait for acks on snapshot iterator done: %w", err)
+		}
+
+		return rec, ErrSnapshotIteratorDone
+	case data := <-s.dataC:
+		s.acks.Add(1)
+		return s.buildRecord(data), nil
+	}
+}
+
+func (s *snapshotIterator) Ack(context.Context, opencdc.Position) error {
+	s.acks.Done()
+	return nil
+}
+
+func (s *snapshotIterator) Teardown(ctx context.Context) error {
+	s.t.Kill(ErrSnapshotIteratorDone)
+	if err := s.t.Err(); err != nil && !errors.Is(err, ErrSnapshotIteratorDone) {
+		return fmt.Errorf(
+			"cannot teardown snapshot mode, fetchers exited unexpectedly: %w", err)
+	}
+
+	if err := s.acks.Wait(ctx); err != nil {
+		return fmt.Errorf("failed to wait for snapshot acks: %w", err)
+	}
+
+	// waiting for the workers to finish will allow us to have an easier time
+	// debugging goroutine leaks.
+	_ = s.t.Wait()
+
+	sdk.Logger(ctx).Info().Msg("all workers done, teared down snapshot iterator")
+
+	return nil
+}
+
+func (s *snapshotIterator) fetchTable(
+	ctx context.Context,
+	tableName common.TableName,
+	primaryKey common.PrimaryKeyName,
+) error {
+	ro := s.config.client.ReadOnlyTransaction()
+	defer ro.Close()
+
+	start, end, err := s.fetchStartEnd(ctx, ro, tableName)
+	if err != nil {
+		return fmt.Errorf("failed to fetch start and end of snapshot: %w", err)
+	}
+
+	query := fmt.Sprint(`
```

Comment:
I somehow though a user will provide the primary key name, but looks like not?

Reply:
Npe, I believe the connector should be smart enought to find the primary key by itself.
I have no idea whether I should validate the given configuration for dangerous parameters though. Is the `pipeline.yml` config file trustable?

Reply:
I agree, it's good if a connector can find the primary key itself. There are some cases where a user wants to specify some custom keys, but that doesn't need to be part of this PR (if we decide to have that feature).

>  I have no idea whether I should validate the given configuration for dangerous parameters though. Is the pipeline.yml config file trustable?

It could happen, as with any user input.:)